app/sidekiq/kms_key_rotation/batch_initiator_job.rb

```diff
@@ -0,0 +1,67 @@
+# frozen_string_literal: true
+
+module KmsKeyRotation
+  class BatchComplete
+    def on_complete(_status, _options)
+      BatchInitiatorJob.perform_async
+    end
+  end
+
+  class BatchInitiatorJob
+    include Sidekiq::Worker
+    include KmsEncryptedModelPatch
+
+    sidekiq_options retry: 5, queue: :low
+
+    RECORDS_PER_BATCH = 10_000
+    RECORDS_PER_JOB = 1_000
+
+    def perform
+      batch = Sidekiq::Batch.new
+      batch.description = "KMS Key Rotation #{batch.bid}"
+      batch.on(:complete, BatchComplete)
+
+      return nil if records.empty?
+
+      batch.jobs do
+        Sidekiq::Client.push_bulk('class' => KmsKeyRotation::RotateKeysJob, 'args' => batched_gids)
+      end
+    end
+
+    def records
+      @records ||= get_records
+    end
+
+    private
+
+    def get_records
+      records = []
+      get_models.each do |model|
+        break if records.size == RECORDS_PER_BATCH
+
+        records += model.where.not('encrypted_kms_key LIKE ? ', "v#{kms_version}:%")
+                        .limit(RECORDS_PER_BATCH - records.size)
+      end
+      records
+    end
+
+    def get_models
+      ApplicationRecord.descendants_using_encryption.map(&:name).map(&:constantize)
+    end
+
+    def batched_gids
+      batched_gids = []
+
+      begin
+        records.map(&:to_global_id)
+               .each_slice(RECORDS_PER_JOB) do |gids|
+                 batched_gids << [{ gids: }.to_json]
```

Comment:
Why do we need to encode this into JSON rather than just passing a hash to Sidekiq?

Reply:
Could we just do:


```ruby
def batched_gids
  records.map(&:to_global_id).each_slice(2).to_a
end
```

And then for the job

```ruby
    def perform(gids)
      Rails.logger.info { "Re-encrypting records: #{gids}" }
      records = GlobalID::Locator.locate_many gids

      skip_hq_callback do
        records.each(&:rotate_kms_key!)
      end
    end
```

Reply:
I was able to remove the JSON encoding, but still had to pass the hash, as Sidekiq::Batch doesn't seem to love passing arrays of arrays as args